File: katas/php/SOLID/lsp/lsp-bad.php
<?php
// Violación del LSP: Las clases derivadas rompen las expectativas de la clase padre
// ❌ Problema: Bird asume que TODOS los pájaros pueden volar

class Bird {
    public function fly(): string {
        return "¡Volando alto en el cielo!";
    }

    public function eat(): string {
        return "Comiendo deliciosa comida";
    }
}

// ❌ Problema: Penguin ES-UN Bird, ¡pero no puede volar!
class Penguin extends Bird {
    public function fly(): string {
        // ❌ Esto viola LSP - rompe la expectativa del padre
        throw new Exception("¡Los pingüinos no pueden volar!"); // ¡Comportamiento roto!
    }

    public function swim(): string {
        return "Nadando con gracia";
    }
}

class Eagle extends Bird {
    public function fly(): string {
        return "¡Volando como un águila!";
    }
}

// Esta función espera que TODOS los pájaros vuelen ❌
function makeBirdFly(Bird $bird): string {
    return $bird->fly(); // ¡Esto lanzará un error para Penguin!
}

// Probando la violación
echo "=== Demostración de Violación LSP ===" . PHP_EOL;

$eagle = new Eagle();
$penguin = new Penguin();

echo "Águila: " . makeBirdFly($eagle) . PHP_EOL; // ✅ Funciona bien

try {
    echo "Pingüino: " . makeBirdFly($penguin) . PHP_EOL; // ❌ ¡SE ROMPE!
} catch (Exception $error) {
    echo "ERROR: " . $error->getMessage() . PHP_EOL;
}

File: katas/python/SOLID/lsp/lsp_bad.py
# Violación del LSP: Las clases derivadas rompen las expectativas de la clase padre
# ❌ Problema: FlyingBird asume que TODOS los pájaros pueden volar


class Bird:
    def fly(self) -> str:
        return "¡Volando alto en el cielo!"

    def eat(self) -> str:
        return "Comiendo deliciosa comida"


# ❌ Problema: Penguin ES-UN Bird, ¡pero no puede volar!
class Penguin(Bird):
    def fly(self) -> str:
        # ❌ Esto viola LSP - rompe la expectativa del padre
        # ¡Comportamiento roto!
        raise Exception("¡Los pingüinos no pueden volar!")

    def swim(self) -> str:
        return "Nadando con gracia"


class Eagle(Bird):
    def fly(self) -> str:
        return "¡Volando como un águila!"


# Esta función espera que TODOS los pájaros vuelen ❌
def make_bird_fly(bird: Bird) -> str:
    return bird.fly()  # ¡Esto lanzará un error para Penguin!


# Probando la violación
if __name__ == "__main__":
    print("=== Demostración de Violación LSP ===")

    eagle = Eagle()
    penguin = Penguin()

    print("Águila:", make_bird_fly(eagle))  # ✅ Funciona bien

    try:
        print("Pingüino:", make_bird_fly(penguin))  # ❌ ¡SE ROMPE!
    except Exception as error:
        print("ERROR:", str(error))
